[WIP] DDF-3700 OpenSearch refactoring and bug fixes

emmberk opened the pull request:
##### DO NOT MERGE. WIP.
____
#### What does this PR do?
This PR includes a few commits to refactor and fix bugs in the OpenSearch Endpoint/Source.

- `Cleaned up catalog-opensearch-source pom` - This fixes fixes the issue of not being able to run unit tests inside the IDE.
- `Moved OpenSearchParser to be a static util class in opensearch-source` - `OpenSearchParser` was originally an interface so that downstream projects could provide their own implementation. However, it makes more sense as a static class instead of a service because all of the methods can be static.
- `Addressed some minor static analysis findings in catalog-opensearch`
- `Added After and Before implementations to OpenSearchFilterVisitor` - This commit adds `Before` and `After` implementations to `OpenSearchFilterVisitor`, instead of dropping those filters from the query. This fixes modified `BEFORE` and `AFTER` temporal queries in Intrigue.
- `Refactored OpenSearchSource so that it does not fail when no keyword is specified -- STILL NEEDS UNIT TESTS` This fixes the issue of failing queries when the keyword was not specified. This commit still needs unit tests.
- `Moved OpenSearch parameter names to new constants class` `opensearch-endpoint` and `opensearch-source` use many of the same constants for query parameters. This commit moves these to a new constants class and also improves the documentation for the source/endpoint.
- `Updated OpenSearchVisitorObject to have getPolygonSearch and getPointRadiusSearch methods instead of getContextualSearch` 
- `Updated polygon to bounding box method in the OpenSearch Source to use com.vividsolutions.jts.geom.Geometry#getEnvelopeInternal` The custom method to construct a bounding box from a polygon was replaced with a third-party method.
- `Added logic in the OpenSearch Source to ignore point-radius filters when the radius is not greater than 0`
- `Added logic in the OpenSearch Source to ignore temporal filters that are not for "modified"`
- `Added logic in the OpenSearch Source to ignore spatial filters that are not for "location"`

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
#### Select relevant component teams: 
@codice/io 
#### Choose 2 committers to review/merge the PR. 
@bdeining
@clockard
#### How should this be tested? (List steps with links to updated documentation)
#### Any background context you want to provide?
#### What are the relevant tickets?
[DDF-3700](https://codice.atlassian.net/browse/DDF-3700)
#### Screenshots (if appropriate)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

clockard commented:
Lots of good things in this pr. Thanks!

stustison commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/query/OpenSearchQuery.java:
```diff
@@ -334,11 +342,11 @@ public void addTypeFilter(String type, String versions) {
       LOGGER.debug("Received versions from client.");
-      String[] typeVersions = versions.split(",");
-      List<Filter> typeVersionPairsFilters = new ArrayList<Filter>();
+      String[] typeVersions = versions.split(OpenSearchParameters.VERSION_DELIMINATOR);
```
Might be more performant to compile the regex from `OpenSearchParameters.VERSION_DELIMINATOR` and then call split from the regex instead.

stustison commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/query/filter/PolygonSpatialFilter.java:
```diff
@@ -14,14 +14,15 @@
     super();
 
-    latLonAry = lonLatStr.split(" |,\\p{Space}?");
+    latLonAry = lonLatStr.split(OpenSearchParameters.POLYGON_LON_LAT_DELIMINATOR);
```
Same comment here about the regex.

stustison commented on catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/query/OsNamespaceContext.java:
```diff
@@ -54,7 +52,7 @@ public String getPrefix(String namespaceURI) {
   @Override
   public Iterator<String> getPrefixes(String namespaceURI) {
-    ArrayList<String> prefixList = new ArrayList<String>(1);
+    ArrayList<String> prefixList = new ArrayList<>(1);
```
Having the size at `1` is going to cause the add to resize the list every single time.
How about just `return Collections.singletonList(getPrefix(namespaceURI)).iterator()`?

bdeining commented on catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchParser.java:
```diff
@@ -41,11 +41,9 @@
 import org.slf4j.LoggerFactory;
 
-public class OpenSearchParserImpl implements OpenSearchParser {
+public final class OpenSearchParser {
```
What's the rationale behind making this final?  What if downstream projects wish to extend this class?

emmberk replied:
I'm open to discussion about this, but my motivation was `OpenSearchParser` only contains some static methods to add some query parameters in a `WebClient`, and it should never be instantiated.

I would think that a new class
```java
public class AnotherProjectOpenSearchParser {

    public static void populateSomething(WebClient client,
```
and using the static methods from each class directly 
```java
AnotherProjectOpenSearchParser.populateSomething(webClient
```
```java
OpenSearchParser.populateSearchOptions(webClient
```
would work well.

bdeining commented on catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/endpoints/OpenSearchEndpointTest.java:
```diff
@@ -114,19 +110,15 @@ public void testProcessQueryForProperHandlingOfSiteNameLOCAL()
+
+              // ***Test verification***
+              // This assert is the whole point of this unit test
+              Assert.assertTrue(
```
ideally we use `assertThat` here.  not sure the comment is useful above

emmberk replied:
There are many of these `Assert` methods in the unit tests in these opensearch bundles that I didn't touch, and it's out of the scope of this ticket to fix all of those.

bdeining commented on catalog/opensearch/catalog-opensearch-common/pom.xml:
```diff
@@ -0,0 +1,37 @@
+            <artifactId>javax.servlet-api</artifactId>
+        </dependency>
+    </dependencies>
+    <parent>
```
✏️  The parent seems to be in a weird location, i would have expected it to be at the top of the pom

emmberk commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/query/filter/BBoxSpatialFilter.java:
```diff
@@ -54,14 +61,29 @@ public BBoxSpatialFilter(String bbox) {
-    wktBuilder.append("," + maxX + " " + minY);
-    wktBuilder.append("," + minX + " " + minY);
-    wktBuilder.append("))");
+    String wktBuilder =
```
It was a static analysis finding to use string concatenation instead of StringBuilder. For readability, I replaced this with 
```java
String.format("POLYGON((%1$s %2$s,%1$s %4$s,%3$s %4$s,%3$s %2$s,%1$s %2$s))", minX, minY, maxX, maxY)
```

mojogitoverhere commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/query/filter/BBoxSpatialFilter.java:
```diff
@@ -30,15 +29,35 @@
     super();
 
-    String[] bboxArY = bbox.split(" |,\\p{Space}?");
+    String[] bboxArY = bbox.split(OpenSearchConstants.BBOX_DELIMITER);
```
❓  Is there a reason the `bbox` string format was made more restrictive?

emmberk replied:
The spec says that the bbox query parameter must be in an exact format:
https://github.com/codice/ddf/blob/3d8784a8d77590b5aaed142d80ea8a68ae4caad0/catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/ddf-os.xml#L59-L60

mojogitoverhere commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/query/filter/BBoxSpatialFilter.java:
```diff
@@ -30,15 +29,35 @@
+              + OpenSearchConstants.BBOX_DELIMITER
+              + "-delimited double values but is \""
+              + bbox
+              + "\".",
```
✏️ `String.format` might be more readable here

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/test/java/ddf/catalog/source/opensearch/impl/OpenSearchFilterVisitorTest.java:
```diff
@@ -112,372 +133,548 @@ public void testAndFilter() {
+  @Test
+  public void testBefore() {
+    org.opengis.filter.temporal.Before beforeFilter =
+        (org.opengis.filter.temporal.Before)
```
It conflicts with the `@org.junit.Before` annotation if it's not qualified.

mojogitoverhere commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/query/filter/PolygonSpatialFilter.java:
```diff
@@ -14,14 +14,15 @@
     super();
 
-    latLonAry = lonLatStr.split(" |,\\p{Space}?");
+    latLonAry = lonLatStr.split(OpenSearchConstants.POLYGON_LON_LAT_DELIMITER);
```
❓Is there a reason the lonLatStr string format was made more restrictive?

emmberk replied:
The spec says that the `polygon` query parameter must be in an exact format:
https://github.com/codice/ddf/blob/3d8784a8d77590b5aaed142d80ea8a68ae4caad0/catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/ddf-os.xml#L62-L64

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchSource.java:
```diff
@@ -301,26 +292,81 @@ public SourceResponse query(QueryRequest queryRequest) throws UnsupportedQueryEx
+      searchPhraseMap = contextualSearch.getSearchPhraseMap();
+    }
+
+    if (!MapUtils.isEmpty(searchPhraseMap)
```
I also fixed a few more that I found.

mackncheesiest commented on catalog/opensearch/catalog-opensearch-api/pom.xml:
```diff
@@ -16,54 +16,27 @@
-
     <parent>
-        <artifactId>catalog-opensearch-pom</artifactId>
+        <artifactId>opensearch</artifactId>
```
✏️ Total nit, but can we make the ordering of groupid : artifactid : version consistent?

emmberk commented:
I think that I've addressed all comments so far. I still need to complete some thorough manual testing, but hopefully that doesn't result in any more code changes.

clockard commented on catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQuery.java:
```diff
@@ -255,11 +273,10 @@ public void addTemporalFilter(String dateStart, String dateEnd, String dateOffse
-    } else if (dateOffset != null && !dateOffset.trim().isEmpty()) {
-      temporalFilter = new TemporalFilter(Long.parseLong(dateOffset));
+    if (StringUtils.isNotBlank(dateStart) || StringUtils.isNotBlank(dateEnd)) {
+      temporalFilter = new TemporalFilter(dateStart.trim(), dateEnd.trim());
```
can dateStart or dateEnd be null here? Just thinking that previously we were checking for null and if one of them isn't null and one is then we would get a NPE on one of the trim calls.

emmberk commented:
build now

peterhuffer commented on catalog/opensearch/catalog-opensearch-api/src/main/java/org/codice/ddf/opensearch/source/PointRadiusSearch.java:
```diff
@@ -38,4 +40,25 @@ public double getLat() {
+      return true;
+    }
+
+    if (!(object instanceof PointRadiusSearch)) {
```
❓ Should we check for null input before doing any other comparisons or does this take care of null here?

emmberk replied:
`instanceOf` already checks for `null`.

clockard commented on catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchFilterVisitor.java:
```diff
@@ -271,16 +269,20 @@ public Object visit(PropertyIsEqualTo filter, Object data) {
+      final Expression expression1 = filter.getExpression1();
+      if (expression1 instanceof PropertyName) {
+        final String propertyName = ((PropertyName) expression1).getPropertyName();
+        final String expectedPropertyIsEqualToTerm = Metacard.ID;
```
Is this local variable just for clarity?

emmberk replied:
Yes. I also use it down below in the log message.

clockard commented on catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchFilterVisitor.java:
```diff
@@ -271,16 +269,20 @@ public Object visit(PropertyIsEqualTo filter, Object data) {
+      if (expression1 instanceof PropertyName) {
+        final String propertyName = ((PropertyName) expression1).getPropertyName();
+        final String expectedPropertyIsEqualToTerm = Metacard.ID;
+        if (StringUtils.equals(expectedPropertyIsEqualToTerm, propertyName)) {
```
Is there a reason we don't just do `expectedPropertyIsEqualToTerm.equals(propertyName)`?

clockard commented on catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java:
```diff
@@ -957,4 +968,78 @@ private WebClient newOpenSearchClient(String url, Subject subj) {
+   * downstream projects may try to implement another algorithm (e.g. best-effort) to combine
+   * searches.
+   *
+   * @return null if the searches cannot be combined, or a {@link SpatialSearch} with one {@link
```
Looks like we return optional and not null

clockard commented on catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java:
```diff
@@ -957,4 +968,78 @@ private WebClient newOpenSearchClient(String url, Subject subj) {
+        CollectionUtils.isNotEmpty(polygonSearches);
+
+    if (filterContainedSomePointRadiusSearchCriteria && filterContainedSomePolygonSearchCriteria) {
+      LOGGER.debug(
```
Hope this doesn't surprise anyone since we are dropping all spatial terms and not taking the last one

emmberk replied:
Right. Before this PR, the last point-radius filter in the query would be used to populate the query parameters. If there were no point-radius filters, then the  last polygon filter in the query would be used to populate the query parameters.
https://github.com/codice/ddf/blob/b279c781e127c5fc9eb73956b9b79609f6b5a5cd/catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchSource.java#L979-L996
The OpenSearch Source was dropping spatial filters without any indication and populating the query parameters from the filters in an unexpected way. This logic was updated to not support any filters with multiple spatial filters. This implementation could be updated in the future or in downstream projects to, for example, combine all spatial criteria into a single `Geometry`.

kcover commented on catalog/opensearch/catalog-opensearch-api/src/main/java/org/codice/ddf/opensearch/source/OpenSearchParser.java:
```diff
@@ -41,9 +41,10 @@ void populateSearchOptions(
-   * OpenSearch specification does not define a syntax for its primary query parameter, searchTerms,
-   * but it is generally used to support simple keyword queries.)
+   * Fills in the OpenSearch query URL with the contextual information is contained in the search
+   * phrase link Map<String, String>}. (Note: Section 2.2 - Query: The OpenSearch specification does
```
✏️ Did you mean to put "information that is contained in the search..." or maybe just "information contained in the search..."?

emmberk commented:
build now

emmberk commented:
build now

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchParserImpl.java:
```diff
@@ -167,41 +144,25 @@ public void populateTemporal(WebClient client, TemporalFilter temporal, List<Str
-    } else {
+    if (geometry != null) {
+      //      checkAndReplace(client, , OpenSearchConstants.GEOMETRY, parameters);
+      // TODO
```
@emmberk reminder to fix this

emmberk commented:
build now

clockard commented:
Successfully ran manual test steps. 👍 :shipit:

emmberk commented:
build now

emmberk commented:
build now
